### DISCUSSION TITLE
launchApp() is deprecated, what is the alternative?

### DISCUSSION THREAD

**Discussion by i-aman0** [appium/appium#19523] [Q&A]
I used driver.launchApp() for launching the app for both android and iOS, but since this is now deprecated, for android I'm using driver.start_activity() but what is the alternative that I can use for iOS?

**Comment by arthuRHD** [comment]
I've used launchApp myself for a while. Documentation has always been unclear about this use case and on why this should be deprecated. I would like to know too 👍.

**Reply by KazuCocoa** [reply]
https://github.com/appium/appium/issues/15807 may help.

> ### Alternative
> - terminate_app and activate_app to restart the app process
>     - Other methods: examples: `mobile: terminateApp`, `mobile: activateApp`, `mobile: launchApp`
> - uninstall/install the app to clean the local data In iOS (iOS does not have clear the local data api, so it requires uninstalling the app to do it)
>     - For iOS simulators, `mobile: clearApp` could help -> https://appium.github.io/appium-xcuitest-driver/latest/execute-methods/#mobile-clearapp
> - https://github.com/appium/appium-uiautomator2-driver#mobile-clearapp to clear the local data in UIA2
> - Espresso driver hasn't supportted resetApp at the beginning



Historically launchApp did re-new session creation internally, it caused unexpected behavior for us historically. So we have introduced activate/terminate app methods to simplify the behavior, just launching the target app's process. `mobile:` commands are also for W3C WebDriver protocol era's extension way.
Some official clients show alternative method descriptions as their deprecation messages as well.

**Comment by berekmerilorand** [comment]
After multiple methods used for a long time were deprecated (`launchApp`, `closeApp` & `background`), I've switched to the below commands.

```
    static async startApp() {
        await driver.activateApp(appPackage);
    }
```

```
    public static async closeApp() {
        await driver.terminateApp(appPackage);
    }
```

```
    private static async reinstallApp() {
        await browser.removeApp(appPackage);
        await browser.pause(Constants.WAIT_TIME_MEDIUM_MS);
        await browser.installApp(appPackage);
    }
```

```
    static async putAppInBackground() {
        driver.isAndroid
            ? await driver.execute('mobile: backgroundApp', appPackage)
            : driver.activateApp('com.apple.Preferences'); # workaround used for iOS because the method used for Android is not working, so I'm opening the Settings app and my testing app is sent to background
```

```
        if (driver.isAndroid) {
            await browser.waitUntil(
                async () => (await browser.queryAppState(appPackage) === 3,
            );
        } else {
            await browser.pause(Constants.WAIT_TIME_LONG_MS);

            await browser.waitUntil(
                async () => (await browser.queryAppState(appPackage)) === 2,
            );
        }
    }
```
